docs/services/api/tapis-v2-v3-migration.md

```diff
@@ -0,0 +1,593 @@
+            }
+          }
+        },
+        {
+          "name": "optionalInput",
+          "description": "Not required, excluded if empty.",
+          "notes": {
+            "details": {
+              "label": "Optional Input File"
+            }
+          }
+        }
+      ]
+    },
+    "fileInputs": [
+      {
+        "name": "requiredInput",
+        "description": "Required input.",
+        "inputMode": "REQUIRED",
+        "targetPath": "*"
+      },
+      {
+        "name": "optionalInput",
+        "description": "Not required, excluded if empty.",
+        "targetPath": "*"
+      },
+      {
+        "name": "fixedInput",
+        "description": "Fixed input.",
+        "inputMode": "FIXED",
+        "sourceUrl": "tapis://data.cyverse.rocks/home/shared/cyverse_training/example/coffee_cake.txt",
+        "targetPath": "*"
+      }
+    ]
+  }
+}
+```
+
+## Tapis v3 Migration Details
+
+### The `notes` Field
+
+Some parameter fields are no longer used in Tapis v3,
+but they can be moved into a `notes` field for use by the DE,
+such as app and parameter `name` and `label` display fields.
+
+The DE will use the app's `notes.label` or `notes.name`,
+followed by the app's `version`,
+when displaying the app name in listings or the in the app launch form.
+If no `label` or `name` is found in the app's `notes`,
+then the DE will simply use the app's `id`.
+
+### Parameters and App Args
+
+Now take the following v2 parameter, for example:
+
+```json
+{
+  "id": "textInput",
+  "value": {
+    "visible": true,
+    "required": false,
+    "type": "string",
+    "default": "default_value"
+  },
+  "details": {
+    "label": "Text Input",
+    "description": "Single-line text."
+  }
+}
+```
+
+This parameter can be defined in v3 `appArgs` like the following:
+
+```json
+{
+  "name": "textInput",
+  "description": "Single-line text.",
+  "arg": "default_value",
+  "notes": {
+    "details": {
+      "label": "Text Input"
+    }
+  }
+}
+```
+
+The `id` field becomes `name`, the `description` moves to the top level,
+and the `value.default` field becomes the `arg` field;
+otherwise the remaining fields move into the `notes` field.
+
+If a `notes.details.label` field is not present,
+then the DE will use the `name` as the parameter's form field label.
+
+Note that the `visible`, `required`, and `type` fields are omitted in this example,
+since those are the defaults the DE will use for those parameter fields.
+
+If the v3 parameter has an `inputMode` field with a `REQUIRED` value,
+then the DE will treat that parameter as required and ignore the
+`notes.value.required` field.
+
+If the v3 parameter has an `inputMode` field with a `FIXED` value,
+then the DE will treat that parameter as hidden and ignore the
+`notes.value.visible` field.
+
+The `notes` object can contain any other custom fields
+(so the entire v2 parameter can be copied as the v3 `notes` field),
+which will be ignored by the DE, with the following exceptions.
```

Comment:
I was expecting exceptions to this rule to appear immediately following this paragraph. Is this referring to the exceptions mentioned above this paragraph?

Reply:
No. I had trouble phrasing this line.

There are exceptions mentioned above this paragraph, and the `notes` field can contain any _other_ custom fields, which will be ignored by the DE, but there are still _other_ fields detailed in the following sections that will not be ignored.

I'll update this line to try and clear up the confusion.